### DISCUSSION TITLE
Shouldn't the ProviderContainer.dispose() be asynchronous?

### DISCUSSION THREAD

**Discussion by dluksza** [rrousselGit/riverpod#823] [Q&A]
What actually happens when `ProviderContainer.dispose()` is called and one of the `onDispose` callbacks is asynchronous?

I guess, since `dispose` is not asynchronous, it won't wait for the `async` disposal callback to finish. _Most_ of the disposable things are actually asynchronous, like canceling stream subscription, closing stream controller, or websocket connection.

Not waiting for the disposal callbacks may theoretically lead to some resources being not fully disposed, or some events (in case of streams) being partially processed, or not processed at all. This is probably less concerning in the case of Flutter app, but it's more of an issue for command-line tools or server applications.

Would it be possible to make `dispose` asynchronous so that we can wait for all of the disposing callbacks to finish before we exit the app?

**Comment by rrousselGit** [comment]
This is a non-issue. Whether `dispose` is asynchronous or not, the Dart VM will still wait for all async operations to complete before quitting the app.

So awaiting `ProviderContainer.dispose` serves no purpose

**Reply by dluksza** [reply]
I'm implementing graceful exit on SIGINT to my command-line app and in this context, it seems that I need to `await` or something.

My signal handler looks like:
```dart
ProcessSignal.sigint.watch().listen((event) {
  container.dispose();
  exit(0);
});
```

I need to call `exit(0)`, because otherwise, the process will hang forever. But this also means that DartVM won't wait for async operations to complete. It will just bailout. This is what I also see in my logs, some of the asynchronous disposal callbacks do not finish. If I add a `Future.delayed(Duration(milliseconds: 500))` between `dispose` and `exit` then everything is disposed correctly.

In this use case having something to await on, before exiting the app would solve the issue.

**Reply by dluksza** [reply]
Out of curiosity, I've added `await container.pump()` between `dispose()` and `exit(0)` and this seems to wait until everything is properly disposed.

**Reply by rrousselGit** [reply]
`pump` is not doing what you want. It's not awaiting for dispose to complete. Rather it seems that your dispose functions complete really fast

**Reply by dluksza** [reply]
Well, they are not that _fast_ to finish within `dispose`, but at least right now _fast_ enough to finish within `pump`.

Anyway, would be good to have something to await for full disposal reliably, as you say the `pump` is just a coincidence.
